web/text_layer_builder.js

```diff
@@ -209,15 +208,70 @@ class TextLayerBuilder {
       }
       end.classList.remove("active");
+      selectingThroughMouse = false;
+      prevRange = null;
+    };
+    div.addEventListener("pointerup", reset);
+
+    // On touchscreen devices, when expaning/reducing a selection through the
+    // 'bubbles' that mark the selection end/start the browser will not fire
+    // any pointer event, because the user is interacting with browser-specific
+    // UI and not with the webpage itself.
+    // To workaround this, we listen to the 'selectionchange' event and check if
+    // the selection includes elements from this page. This has two drawbacks:
+    //  - We can only detect changes once they start happening. If the first
+    //    movement that the user does is to move one of the selection bubbles
+    //    to a blank part of the page, the selection will flicker before
+    //    that we can apply the fix.
+    //  - We cannot detect when the user stops interacting with the selection,
+    //    because there is no difference between "the user is still holding
+    //    their finger on the selection 'bubble'" and "the user is lifter their
+    //    finger and then pressed the 'bubble' again". Thus, the .endOfContent
+    //    element will remain active until when the user interacts with the
+    //    page again, triggering a pointerup event.
+    document.addEventListener("selectionchange", () => {
+      if (selectingThroughMouse) return;
+
+      const selection = document.getSelection();
+      if (selection.rangeCount === 0) return reset();
+
+      const range = selection.getRangeAt(0);
+      if (!range.intersectsNode(div)) return reset();
+
+      if (typeof PDFJSDev === "undefined" || !PDFJSDev.test("MOZCENTRAL")) {
+        isFirefox ??=
+          (typeof PDFJSDev === "undefined" || !PDFJSDev.test("CHROME")) &&
+          getComputedStyle(end).getPropertyValue("-moz-user-select") === "none";
+        // In non-Firefox browsers, when hovering over an empty space (thus, on
+        // .endOfContent), the selection will expand to cover all the text between
+        // the current selection and .endOfContent. By moving .endOfContent to
+        // right after (or before, depending on which side of the selection the user
+        // is moving), we limit the selection jump to at most cover the enteirety
+        // of the <span> where the selection is being modified.
+        if (!isFirefox) {
+          const modifyStart =
+            prevRange &&
+            (range.compareBoundaryPoints(Range.END_TO_END, prevRange) === 0 ||
+              range.compareBoundaryPoints(Range.START_TO_END, prevRange) === 0);
+          let anchor = modifyStart ? range.startContainer : range.endContainer;
+          if (anchor.nodeType === Node.TEXT_NODE) anchor = anchor.parentNode;
+          // Make sure that we are always only moving .endOfContent to somewhere
+          // directly in .textLayer, and not inside nested spans.
+          anchor = anchor.closest(".textLayer > span");
```

Comment:
The behavior in Chrome is still not perfect with PDFs that use `.markedContent`: this PR moves the `.endOfContent` before/after `.markedContent`, so all the `<span>s` inside `.markedContent` are considered as a "single block" and the selection will flicker betwen the "correct selection" and selecting the whole `.markedContent` block. I will eventually try to fix it, but for now this is already an improvement compared to the current "flicker over the whole page" behavior.